ColumnMetaDataLoader.load throw exception on ShardingSphereDataSource

## Bug Report

### Which version of ShardingSphere did you use?
master branch, commit: 3b8dc82bfb8691f52f221ae394587bcfc03bfa9b

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy, scaling

### Expected behavior
Load column metadata successfully.

It should be a new scernario of `ColumnMetaDataLoader.load` on ShardingSphereDataSource, which is used by scailng data consistency check to get primary key from metadata.

### Actual behavior
Exception thrown:
```
[_finished_check_Worker-1] o.a.s.d.p.core.job.FinishedCheckJob - scaling job 5011358192994937 finish check failed!
org.apache.shardingsphere.data.pipeline.core.exception.DataCheckFailException: get table columns failed
        at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyCheckerImpl.getTablesColumnsMap(DataConsistencyCheckerImpl.java:198)
        at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyCheckerImpl.checkRecordsContent(DataConsistencyCheckerImpl.java:130)
        at org.apache.shardingsphere.data.pipeline.core.api.impl.PipelineJobAPIImpl.dataConsistencyCheck0(PipelineJobAPIImpl.java:253)
        at org.apache.shardingsphere.data.pipeline.core.api.impl.PipelineJobAPIImpl.dataConsistencyCheck(PipelineJobAPIImpl.java:235)
        at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.dataConsistencyCheck(FinishedCheckJob.java:100)
        at org.apache.shardingsphere.data.pipeline.core.job.FinishedCheckJob.execute(FinishedCheckJob.java:79)
        at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:33)
        at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:29)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:172)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:141)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:124)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:100)
        at org.apache.shardingsphere.elasticjob.lite.internal.schedule.LiteJob.execute(LiteJob.java:35)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.sql.SQLException: Column index out of range.
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63)
        at com.mysql.cj.jdbc.result.ResultSetMetaData.getField(ResultSetMetaData.java:223)
        at com.mysql.cj.jdbc.result.ResultSetMetaData.isAutoIncrement(ResultSetMetaData.java:277)
        at org.apache.shardingsphere.driver.jdbc.core.resultset.ShardingSphereResultSetMetaData.isAutoIncrement(ShardingSphereResultSetMetaData.java:62)
        at org.apache.shardingsphere.infra.metadata.schema.builder.loader.common.ColumnMetaDataLoader.load(ColumnMetaDataLoader.java:77)
        at org.apache.shardingsphere.infra.metadata.schema.builder.loader.common.TableMetaDataLoader.load(TableMetaDataLoader.java:51)
        at org.apache.shardingsphere.data.pipeline.core.check.consistency.DataConsistencyCheckerImpl.getTablesColumnsMap(DataConsistencyCheckerImpl.java:192)
        ... 14 common frames omitted
```

### Reason analyze (If you can)
1) `ShardingSphereDatabaseMetaData.getActualTableNamePattern` expand `tableNamePattern` as `% tableName %`, e.g. `t_order` will be changed to `%t_order%`. When it's executed in underlying DB, it'll get sharded tables columns metadata, e.g. t_order_0, t_order_1, etc.

2) `ColumnMetaDataLoader.load` doesn't isolate `columnNames` by table name, they're put together into List<String>.

<img width="1371" alt="图片" src="https://user-images.githubusercontent.com/42492540/147644140-3713ac1e-e877-4df5-bdee-6cd2fd19db2c.png">


### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java
@@ -27,14 +27,14 @@ import org.apache.shardingsphere.data.pipeline.api.datasource.config.PipelineDat
 import org.apache.shardingsphere.data.pipeline.api.datasource.config.PipelineDataSourceConfigurationFactory;
 import org.apache.shardingsphere.data.pipeline.core.datasource.PipelineDataSourceFactory;
 import org.apache.shardingsphere.data.pipeline.core.exception.PipelineDataConsistencyCheckFailedException;
+import org.apache.shardingsphere.data.pipeline.core.metadata.loader.PipelineTableMetaDataLoader;
+import org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaData;
 import org.apache.shardingsphere.data.pipeline.scenario.rulealtered.RuleAlteredJobContext;
 import org.apache.shardingsphere.data.pipeline.spi.check.consistency.DataConsistencyCheckAlgorithm;
 import org.apache.shardingsphere.data.pipeline.spi.check.consistency.SingleTableDataCalculator;
 import org.apache.shardingsphere.data.pipeline.spi.ratelimit.JobRateLimitAlgorithm;
 import org.apache.shardingsphere.infra.database.type.DatabaseType;
 import org.apache.shardingsphere.infra.executor.kernel.thread.ExecutorThreadFactoryBuilder;
-import org.apache.shardingsphere.infra.metadata.schema.builder.loader.common.TableMetaDataLoader;
-import org.apache.shardingsphere.infra.metadata.schema.model.TableMetaData;
 import org.apache.shardingsphere.scaling.core.job.sqlbuilder.ScalingSQLBuilderFactory;
 
 import javax.sql.DataSource;
@@ -48,7 +48,6 @@ import java.util.Iterator;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
@@ -127,13 +126,6 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
                 jobContext.getJobConfig().getPipelineConfig().getTarget().getType(), jobContext.getJobConfig().getPipelineConfig().getTarget().getParameter());
         checkDatabaseTypeSupportedOrNot(supportedDatabaseTypes, targetDataSourceConfig.getDatabaseType().getName());
         Collection<String> logicTableNames = jobContext.getTaskConfigs().stream().flatMap(each -> each.getDumperConfig().getTableNameMap().values().stream()).distinct().collect(Collectors.toList());
-        Map<String, TableMetaData> tableMetaDataMap = getTablesColumnsMap(sourceDataSourceConfig, logicTableNames);
-        logicTableNames.forEach(each -> {
-            //TODO put to preparer
-            if (!tableMetaDataMap.containsKey(each)) {
-                throw new PipelineDataConsistencyCheckFailedException(String.format("could not get columns for table '%s'", each));
-            }
-        });
         String sourceDatabaseType = sourceDataSourceConfig.getDatabaseType().getName();
         String targetDatabaseType = targetDataSourceConfig.getDatabaseType().getName();
         SingleTableDataCalculator sourceCalculator = checkAlgorithm.getSingleTableDataCalculator(sourceDatabaseType);
@@ -144,9 +136,17 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
         JobRateLimitAlgorithm rateLimitAlgorithm = jobContext.getRuleAlteredContext().getRateLimitAlgorithm();
         try (PipelineDataSourceWrapper sourceDataSource = dataSourceFactory.newInstance(sourceDataSourceConfig);
              PipelineDataSourceWrapper targetDataSource = dataSourceFactory.newInstance(targetDataSourceConfig)) {
+            PipelineTableMetaDataLoader tableMetaDataLoader = getTableMetaDataLoader(sourceDataSource);
+            logicTableNames.forEach(each -> {
+                //TODO put to preparer
+                if (null == tableMetaDataLoader.getTableMetaData(each)) {
+                    throw new PipelineDataConsistencyCheckFailedException(String.format("could not get metadata for table '%s'", each));
+                }
+            });
             for (String each : logicTableNames) {
-                Collection<String> columnNames = tableMetaDataMap.get(each).getColumns().keySet();
-                String uniqueKey = tableMetaDataMap.get(each).getPrimaryKeyColumns().get(0);
+                PipelineTableMetaData tableMetaData = tableMetaDataLoader.getTableMetaData(each);
+                Collection<String> columnNames = tableMetaData.getColumnNames();
+                String uniqueKey = tableMetaData.getPrimaryKeys().get(0);
                 DataCalculateParameter sourceCalculateParameter = DataCalculateParameter.builder().dataSource(sourceDataSource).databaseType(sourceDatabaseType).peerDatabaseType(targetDatabaseType)
                     .logicTableName(each).columnNames(columnNames).uniqueKey(uniqueKey).build();
                 DataCalculateParameter targetCalculateParameter = DataCalculateParameter.builder().dataSource(targetDataSource).databaseType(targetDatabaseType).peerDatabaseType(sourceDatabaseType)
@@ -170,7 +170,7 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
                 result.put(each, calculateResultsEquals);
             }
         } catch (final ExecutionException | InterruptedException | SQLException ex) {
-            throw new PipelineDataConsistencyCheckFailedException("data check failed");
+            throw new PipelineDataConsistencyCheckFailedException("data check failed", ex);
         } finally {
             executor.shutdown();
             executor.shutdownNow();
@@ -185,22 +185,9 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
     }
     
     // TODO reuse metadata
-    private Map<String, TableMetaData> getTablesColumnsMap(final PipelineDataSourceConfiguration dataSourceConfig, final Collection<String> tableNames) {
-        try (PipelineDataSourceWrapper dataSource = dataSourceFactory.newInstance(dataSourceConfig)) {
-            Map<String, TableMetaData> result = new LinkedHashMap<>();
-            for (String each : tableNames) {
-                Optional<TableMetaData> tableMetaDataOptional;
-                try {
-                    tableMetaDataOptional = TableMetaDataLoader.load(dataSource, each, dataSourceConfig.getDatabaseType());
-                } catch (final SQLException ex) {
-                    throw new PipelineDataConsistencyCheckFailedException(String.format("get columns failed for table '%s'", each), ex);
-                }
-                TableMetaData tableMetaData = tableMetaDataOptional.orElseThrow(() -> new PipelineDataConsistencyCheckFailedException(String.format("get metadata failed for table '%s'", each)));
-                result.put(each, tableMetaData);
-            }
-            return result;
-        } catch (final SQLException ex) {
-            throw new PipelineDataConsistencyCheckFailedException("create data source failed", ex);
+    private PipelineTableMetaDataLoader getTableMetaDataLoader(final PipelineDataSourceWrapper sourceDataSource) throws SQLException {
+        try (Connection connection = sourceDataSource.getConnection()) {
+            return new PipelineTableMetaDataLoader(connection, "%");
         }
     }
 }

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/DataConsistencyCheckerImpl.java
@@ -49,6 +49,7 @@ import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.ThreadFactory;
@@ -64,6 +65,8 @@ import java.util.stream.Collectors;
 @Getter
 @Slf4j
 public final class DataConsistencyCheckerImpl implements DataConsistencyChecker {
+    
+    private static final Map<PipelineDataSourceConfiguration, PipelineTableMetaDataLoader> TABLE_META_DATA_LOADER_MAP = new ConcurrentHashMap<>();
     
     private final PipelineDataSourceFactory dataSourceFactory = new PipelineDataSourceFactory();
     
@@ -136,7 +139,7 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
         JobRateLimitAlgorithm rateLimitAlgorithm = jobContext.getRuleAlteredContext().getRateLimitAlgorithm();
         try (PipelineDataSourceWrapper sourceDataSource = dataSourceFactory.newInstance(sourceDataSourceConfig);
              PipelineDataSourceWrapper targetDataSource = dataSourceFactory.newInstance(targetDataSourceConfig)) {
-            PipelineTableMetaDataLoader tableMetaDataLoader = getTableMetaDataLoader(sourceDataSource);
+            PipelineTableMetaDataLoader tableMetaDataLoader = getTableMetaDataLoader(sourceDataSourceConfig, sourceDataSource);
             logicTableNames.forEach(each -> {
                 //TODO put to preparer
                 if (null == tableMetaDataLoader.getTableMetaData(each)) {
@@ -184,10 +187,21 @@ public final class DataConsistencyCheckerImpl implements DataConsistencyChecker 
         }
     }
     
-    // TODO reuse metadata
-    private PipelineTableMetaDataLoader getTableMetaDataLoader(final PipelineDataSourceWrapper sourceDataSource) throws SQLException {
-        try (Connection connection = sourceDataSource.getConnection()) {
-            return new PipelineTableMetaDataLoader(connection, "%");
+    private PipelineTableMetaDataLoader getTableMetaDataLoader(final PipelineDataSourceConfiguration sourceDataSourceConfig, final PipelineDataSourceWrapper sourceDataSource) throws SQLException {
+        PipelineTableMetaDataLoader result = TABLE_META_DATA_LOADER_MAP.get(sourceDataSourceConfig);
+        if (null != result) {
+            return result;
+        }
+        synchronized (TABLE_META_DATA_LOADER_MAP) {
+            result = TABLE_META_DATA_LOADER_MAP.get(sourceDataSourceConfig);
+            if (null != result) {
+                return result;
+            }
+            try (Connection connection = sourceDataSource.getConnection()) {
+                result = new PipelineTableMetaDataLoader(connection, "%");
+                TABLE_META_DATA_LOADER_MAP.put(sourceDataSourceConfig, result);
+            }
+            return result;
         }
     }
 }

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/loader/PipelineTableMetaDataLoader.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/loader/PipelineTableMetaDataLoader.java
@@ -1,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.data.pipeline.core.metadata.loader;
+
+import org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineColumnMetaData;
+import org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaData;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+/**
+ * Pipeline table meta data loader.
+ */
+public final class PipelineTableMetaDataLoader {
+    
+    private final Map<String, PipelineTableMetaData> tableMetaDataMap;
+    
+    public PipelineTableMetaDataLoader(final Connection connection, final String tableNamePattern) throws SQLException {
+        this.tableMetaDataMap = loadTableMetadataMap(connection, tableNamePattern);
+    }
+    
+    private Map<String, PipelineTableMetaData> loadTableMetadataMap(final Connection connection, final String tableNamePattern) throws SQLException {
+        Map<String, Map<String, PipelineColumnMetaData>> tablePipelineColumnMetaDataMap = new LinkedHashMap<>();
+        Map<String, Set<String>> primaryKeys = loadPrimaryKeys(connection, tableNamePattern);
+        try (ResultSet resultSet = connection.getMetaData().getColumns(connection.getCatalog(), connection.getSchema(), tableNamePattern, "%")) {
+            while (resultSet.next()) {
+                String tableName = resultSet.getString("TABLE_NAME");
+                Map<String, PipelineColumnMetaData> columnMetaDataMap = tablePipelineColumnMetaDataMap.computeIfAbsent(tableName, k -> new LinkedHashMap<>());
+                String columnName = resultSet.getString("COLUMN_NAME");
+                if (columnMetaDataMap.containsKey(columnName)) {
+                    continue;
+                }
+                int dataType = resultSet.getInt("DATA_TYPE");
+                boolean primaryKey = primaryKeys.containsKey(tableName) && primaryKeys.get(tableName).contains(columnName);
+                PipelineColumnMetaData columnMetaData = new PipelineColumnMetaData(columnName, dataType, primaryKey);
+                columnMetaDataMap.put(columnName, columnMetaData);
+            }
+        }
+        Map<String, PipelineTableMetaData> result = new LinkedHashMap<>();
+        for (Entry<String, Map<String, PipelineColumnMetaData>> entry : tablePipelineColumnMetaDataMap.entrySet()) {
+            result.put(entry.getKey(), new PipelineTableMetaData(entry.getKey(), entry.getValue()));
+        }
+        return result;
+    }
+    
+    private Map<String, Set<String>> loadPrimaryKeys(final Connection connection, final String tableNamePattern) throws SQLException {
+        Map<String, Set<String>> result = new LinkedHashMap<>();
+        try (ResultSet resultSet = connection.getMetaData().getPrimaryKeys(connection.getCatalog(), connection.getSchema(), tableNamePattern)) {
+            while (resultSet.next()) {
+                String tableName = resultSet.getString("TABLE_NAME");
+                Set<String> columnNames = result.computeIfAbsent(tableName, k -> new LinkedHashSet<>());
+                columnNames.add(resultSet.getString("COLUMN_NAME"));
+            }
+        }
+        return result;
+    }
+    
+    /**
+     * Get table metadata.
+     *
+     * @param tableName table name
+     * @return table metadata
+     */
+    public PipelineTableMetaData getTableMetaData(final String tableName) {
+        return tableMetaDataMap.get(tableName);
+    }
+}

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/loader/PipelineTableMetaDataLoader.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/loader/PipelineTableMetaDataLoader.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.data.pipeline.core.metadata.loader;
 
+import lombok.extern.slf4j.Slf4j;
 import org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineColumnMetaData;
 import org.apache.shardingsphere.data.pipeline.core.metadata.model.PipelineTableMetaData;
 
@@ -32,6 +33,7 @@ import java.util.Set;
 /**
  * Pipeline table meta data loader.
  */
+@Slf4j
 public final class PipelineTableMetaDataLoader {
     
     private final Map<String, PipelineTableMetaData> tableMetaDataMap;
@@ -42,8 +44,8 @@ public final class PipelineTableMetaDataLoader {
     
     private Map<String, PipelineTableMetaData> loadTableMetadataMap(final Connection connection, final String tableNamePattern) throws SQLException {
         Map<String, Map<String, PipelineColumnMetaData>> tablePipelineColumnMetaDataMap = new LinkedHashMap<>();
-        Map<String, Set<String>> primaryKeys = loadPrimaryKeys(connection, tableNamePattern);
-        try (ResultSet resultSet = connection.getMetaData().getColumns(connection.getCatalog(), connection.getSchema(), tableNamePattern, "%")) {
+        // TODO if tableNamePattern is '%', it might return inconsistent result, actual table `t_order_2` may be return
+        try (ResultSet resultSet = connection.getMetaData().getColumns(connection.getCatalog(), null, tableNamePattern, "%")) {
             while (resultSet.next()) {
                 String tableName = resultSet.getString("TABLE_NAME");
                 Map<String, PipelineColumnMetaData> columnMetaDataMap = tablePipelineColumnMetaDataMap.computeIfAbsent(tableName, k -> new LinkedHashMap<>());
@@ -52,7 +54,14 @@ public final class PipelineTableMetaDataLoader {
                     continue;
                 }
                 int dataType = resultSet.getInt("DATA_TYPE");
-                boolean primaryKey = primaryKeys.containsKey(tableName) && primaryKeys.get(tableName).contains(columnName);
+                Set<String> primaryKeys;
+                try {
+                    primaryKeys = loadPrimaryKeys(connection, tableName);
+                } catch (final SQLException ex) {
+                    log.error("loadPrimaryKeys failed, tableName={}", tableName);
+                    throw ex;
+                }
+                boolean primaryKey = primaryKeys.contains(columnName);
                 PipelineColumnMetaData columnMetaData = new PipelineColumnMetaData(columnName, dataType, primaryKey);
                 columnMetaDataMap.put(columnName, columnMetaData);
             }
@@ -61,16 +70,15 @@ public final class PipelineTableMetaDataLoader {
         for (Entry<String, Map<String, PipelineColumnMetaData>> entry : tablePipelineColumnMetaDataMap.entrySet()) {
             result.put(entry.getKey(), new PipelineTableMetaData(entry.getKey(), entry.getValue()));
         }
+        log.info("loadTableMetadataMap, result={}", result);
         return result;
     }
     
-    private Map<String, Set<String>> loadPrimaryKeys(final Connection connection, final String tableNamePattern) throws SQLException {
-        Map<String, Set<String>> result = new LinkedHashMap<>();
-        try (ResultSet resultSet = connection.getMetaData().getPrimaryKeys(connection.getCatalog(), connection.getSchema(), tableNamePattern)) {
+    private Set<String> loadPrimaryKeys(final Connection connection, final String tableName) throws SQLException {
+        Set<String> result = new LinkedHashSet<>();
+        try (ResultSet resultSet = connection.getMetaData().getPrimaryKeys(connection.getCatalog(), connection.getSchema(), tableName)) {
             while (resultSet.next()) {
-                String tableName = resultSet.getString("TABLE_NAME");
-                Set<String> columnNames = result.computeIfAbsent(tableName, k -> new LinkedHashSet<>());
-                columnNames.add(resultSet.getString("COLUMN_NAME"));
+                result.add(resultSet.getString("COLUMN_NAME"));
             }
         }
         return result;

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineColumnMetaData.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineColumnMetaData.java
@@ -1,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.data.pipeline.core.metadata.model;
+
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+import java.util.Objects;
+
+/**
+ * Column meta data.
+ */
+@RequiredArgsConstructor
+@Getter
+@ToString
+public final class PipelineColumnMetaData {
+    
+    @NonNull
+    private final String name;
+    
+    private final int dataType;
+    
+    private final boolean primaryKey;
+    
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final PipelineColumnMetaData that = (PipelineColumnMetaData) o;
+        return name.equals(that.name);
+    }
+    
+    @Override
+    public int hashCode() {
+        return Objects.hash(name);
+    }
+}

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaData.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaData.java
@@ -1,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.data.pipeline.core.metadata.model;
+
+import lombok.Getter;
+import lombok.ToString;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Pipelien table meta data.
+ */
+@ToString
+public final class PipelineTableMetaData {
+    
+    private final String name;
+    
+    private final Map<String, PipelineColumnMetaData> columnMetaDataMap;
+    
+    @Getter
+    private final Collection<String> columnNames;
+    
+    @Getter
+    private final List<String> primaryKeys;
+    
+    public PipelineTableMetaData(final String name, final Map<String, PipelineColumnMetaData> columnMetaDataMap) {
+        this.name = name;
+        this.columnMetaDataMap = columnMetaDataMap;
+        columnNames = Collections.unmodifiableCollection(columnMetaDataMap.keySet());
+        primaryKeys = Collections.unmodifiableList(columnMetaDataMap.values().stream().filter(PipelineColumnMetaData::isPrimaryKey).map(PipelineColumnMetaData::getName).collect(Collectors.toList()));
+    }
+}

--- a/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaData.java
+++ b/shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/metadata/model/PipelineTableMetaData.java
@@ -18,12 +18,14 @@
 package org.apache.shardingsphere.data.pipeline.core.metadata.model;
 
 import lombok.Getter;
+import lombok.NonNull;
 import lombok.ToString;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
 /**
@@ -32,6 +34,7 @@ import java.util.stream.Collectors;
 @ToString
 public final class PipelineTableMetaData {
     
+    @NonNull
     private final String name;
     
     private final Map<String, PipelineColumnMetaData> columnMetaDataMap;
@@ -47,5 +50,22 @@ public final class PipelineTableMetaData {
         this.columnMetaDataMap = columnMetaDataMap;
         columnNames = Collections.unmodifiableCollection(columnMetaDataMap.keySet());
         primaryKeys = Collections.unmodifiableList(columnMetaDataMap.values().stream().filter(PipelineColumnMetaData::isPrimaryKey).map(PipelineColumnMetaData::getName).collect(Collectors.toList()));
+    }
+    
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final PipelineTableMetaData that = (PipelineTableMetaData) o;
+        return name.equals(that.name);
+    }
+    
+    @Override
+    public int hashCode() {
+        return Objects.hash(name);
     }
 }

